### PR TITLE
New tools and bug fixes

### PR DESCRIPTION
**Pull request by QU35T-code**
# Description

This pull request contains 16 new tools to the Exegol image. Problems encountered on Issue #72 and Issue #78  are also fixed here.

# New tools

## Forensics

* volatility3
* testdisk
* fdisk
* sleuthkit
* objection

## Steganography

* exif
* exiv2
* hexedit

## Web

* php_filter_chain_generator
* shuffledns
* tailscale
* swaks
* dnsx

## Reverse

* jadx

## Code Analysis

* brakeman
* semgrep

## History

* brakeman
* dnsx
* exif
* exiv2
* fdisk
* hexedit
* jadx
* objection
* semgrep
* shuffledns
* subfinder
* swaks
* tailscale
* volatility3
* testdisk

## Aliases

* jadx
* ligolo-ng
* php_filter_chain_generator
* tailscale
* volatility3

# Fixes

* Fix **tor** dependency with **missing -y flag**
* Fix Issue #72
* Fix Issue #78
* mdcat install
* maigret install
* crackmapexec install

### CODE REVIEW

**Comment by ShutdownRepo** [comment]
Hi @QU35T-code ! Thank you for those additions! The problem is we are in the middle of important changes and refactoring. We will be releasing those changes very soon and you should then be able to update your PR accordingly.

Briefly put, the changes needed here will be
- add a test command that our new CI/CD pipeline will be able to run to make sure that tools are installed correctly, something like `tool --help`.
- the history and alias system will be a bit different. If `toolA` needs a set of aliases and history commands, new files must be created in `/sources/zsh/alias.d` and `/sources/zsh/history.d`. Each install function will then be able to call the `add-aliases` and/or `add-history` functions and point to those files. This will allow history and aliases to be "installed" only if the corresponding tool is installed.
- a comment will also need to be added in order to explaining briefly what the tool does 😉 

Just a quick question though, why do you fix python3 version to 3.9 ?

**Comment by QU35T-code** [comment]
Hey @ShutdownRepo !

Message received, I will make the changes quickly !

Concerning python, it is especially for the installation of volatility3 where there was a problem.
Some functions no longer seem to work under python3.10 and I wondered if the same thing happened with other tools.
Currently, most of the tools have been developed under python3.10 and may conflict with python3.10.
That's why I think it would be best to have the latest version of python3 but also python3.9.

**Note:** Currently in my PR, python3.9 is installed directly through the package, but wouldn't it be better to do it through the package manager ? What do you think of this ;) ?

**Comment by ShutdownRepo** [comment]
Okay understood, makes sense. Installing python3.9 through the package manager would be better in my opinion, we decrease the risk of failing during the install

**Comment by QU35T-code** [comment]
I did not understand where I have to add the command to trigger the CI / CD ?

**Comment by Dramelac** [comment]
We come back to you after the release of the new version to explain you everything in detail.

**Comment by QU35T-code** [comment]
Hello @ShutdownRepo and @Dramelac,

I made the necessary changes so that my PR no longer has any conflicts.
Finally, my addition of python3.9 was removed because the Debian repository has `Python 3.9` as the default version (not `Python 3.10`).
I added the `aliases`, `history` and `test command` in the `install.sh` script.
If you have any other comments, please do not hesitate.

Regards,

File: sources/zsh/aliases
Comment (Dramelac) [inline sources/zsh/aliases:17]: Concerning the `www` alias, it would be better not to specify the port here to allow the user to change it if needed while using the alias, for example: `www 8080`

**Review by Dramelac — CHANGES_REQUESTED** [review]
First quick review, to confirm with @ShutdownRepo if we have to use pipx.

File: sources/install.sh
Comment (QU35T-code) [reply sources/install.sh:2310]: It is true that it is better not to launch the service when it builds, but only when necessary !
I move the command to tailscale history 👍

File: sources/install.sh
Comment (QU35T-code) [reply sources/install.sh:2323]: Yep, I can get the helper from compiled binaries ! 👍

File: sources/zsh/aliases
Comment (QU35T-code) [reply sources/zsh/aliases:17]: Good idea ! 👍

File: sources/install.sh
Comment (QU35T-code) [reply sources/install.sh:2100]: I had used this specific version because I had a dependency problem on the last one.
To solve the problem, I now install objection with `pipx` ! 👍

File: sources/install.sh
Comment (QU35T-code) [reply sources/install.sh]: I solved the problem ! #pipx4thewin

**Comment by QU35T-code** [comment]
Hey @Dramelac,

I solved the problems of the review on this PR. I'll let you come back to me if you have any other comments ! 👍

File: sources/install.sh
Comment (Dramelac) [reply sources/install.sh]: The daemon command is still here (line 2305), is there a reason why ?

File: sources/install.sh
Comment (Dramelac) [inline sources/install.sh:2746]: Can you use `fapt` here or `apt-get`instead of `apt` ?

File: sources/install.sh
Comment (Dramelac) [inline sources/install.sh:2304]: Can you use `fapt` here or `apt-get` instead of `apt` ?

File: sources/zsh/aliases
Comment (Dramelac) [inline sources/zsh/aliases:18]: To check because i'm not sure but i don't think the `$1` is necessary, just `alias www="python3 -m http.server"` should be enough.

File: sources/zsh/aliases.d/ligolo-ng
Comment (Dramelac) [inline sources/zsh/aliases.d/ligolo-ng:1]: There is a missing new line at the end of the file

File: sources/zsh/history.d/tailscale
Comment (Dramelac) [inline sources/zsh/history.d/tailscale:2]: Missing new line here too at the end of the last line

**Review by Dramelac — CHANGES_REQUESTED** [review]
Almost there, thank you for your help and all your work in this PR! 
Last details and i think we are ready to merge

File: sources/install.sh
Comment (QU35T-code) [reply sources/install.sh]: It's a mistake !

File: sources/zsh/aliases.d/ligolo-ng
Comment (ShutdownRepo) [reply sources/zsh/aliases.d/ligolo-ng:1]: Should be okay as-is, we have a function that cleans things up

File: sources/zsh/history.d/tailscale
Comment (ShutdownRepo) [reply sources/zsh/history.d/tailscale:2]: Should be okay as-is, we have a function that cleans things up

**Review by Dramelac — APPROVED** [review]

**Comment by Dramelac** [comment]
OK for me, thank you @QU35T-code for this work!
@ShutdownRepo I'll let you run the merge if you're comfortable running it without having done a CI/CD build and test before or otherwise we'll wait to improve / fix the PR pipeline.

File: sources/zsh/history.d/volatility3
Comment (ShutdownRepo) [inline sources/zsh/history.d/volatility3:15]: I think in day-to-day operations, users won't create the `$DUMP` environment variable. What do you think about replacing `$DUMP` with `memory.dmp`?

**Review by ShutdownRepo — CHANGES_REQUESTED** [review]

File: sources/zsh/history.d/volatility3
Comment (QU35T-code) [reply sources/zsh/history.d/volatility3:15]: Yes I agree, it's more intuitive !

**Comment by QU35T-code** [comment]
Hello @ShutdownRepo 

Thank you for the feedback. I think I made the necessary changes that are appropriate !
I also added new commands which are useful especially for forensics and steganography (Sorry, maybe we'll merge one day ^^) :

```bash
* testdisk
* jadx
* fdisk
* exif
* exiv2
* hexedit
* sleuthkit
```

I let you review all this with @Dramelac :)

**Comment by QU35T-code** [comment]
Ups, I didn't know I couldn't request for review you 2 at the same time.

**Review by Dramelac — APPROVED** [review]
The code looks good to me, ideally we could also use this RP to test CI/CD pipelines in an RP context but it is still a WIP.
I validate the PR on my side, @ShutdownRepo I let you make a pass to see if we merge right away or we wait until the test pipelines for PR are ready.

File: sources/install.sh
Comment (ShutdownRepo) [inline sources/install.sh:1977]: Fyi you don't necessarily have to define lots of test commands. The purpose of test commands is to make sure the tool or framework is installed properly. In this particular case we can keep them, but if one or two commands work, there's no reason the others wouldn't.

File: sources/install.sh
Comment (ShutdownRepo) [inline sources/install.sh:2324]: Just for the sake of uniformity, could you switch to `python3 -m pip install` please?

**Review by ShutdownRepo — CHANGES_REQUESTED** [review]
One last change and we're good to go 😉

**Comment by ShutdownRepo** [comment]
> The code looks good to me, ideally we could also use this RP to test CI/CD pipelines in an RP context but it is still a WIP. I validate the PR on my side, @ShutdownRepo I let you make a pass to see if we merge right away or we wait until the test pipelines for PR are ready.

The PR will be ready for merge soon. In my opinion, whether we merge it now or use it to qualify our CI/CD pipeline, we shouldn't wait too long before doing so. We'll seen how our tests and building of the pipeline progress, and we'll decide accordingly. Whatever happens, let's have a deadline for this PR: Feb. 28th

**Comment by QU35T-code** [comment]
> > The code looks good to me, ideally we could also use this RP to test CI/CD pipelines in an RP context but it is still a WIP. I validate the PR on my side, @ShutdownRepo I let you make a pass to see if we merge right away or we wait until the test pipelines for PR are ready.
> 
> The PR will be ready for merge soon. In my opinion, whether we merge it now or use it to qualify our CI/CD pipeline, we shouldn't wait too long before doing so. We'll seen how our tests and building of the pipeline progress, and we'll decide accordingly. Whatever happens, let's have a deadline for this PR: Feb. 28th

No problem. Sure, it's better if the pipeline is ready for the merge !

**Review by ShutdownRepo — APPROVED** [review]

**Comment by ShutdownRepo** [comment]
Congrats on this PR, you did some very nice work @QU35T-code !!
We'll merge before the deadline

**Comment by ShutdownRepo** [comment]
@QU35T-code would you mind editing this PR's first comment so that it reflects its final state? Will be easier when merging, for the release message etc.

**Comment by psyray** [comment]
Awesome work, thanks @QU35T-code

**Comment by psyray** [comment]
@ShutdownRepo 
Shouldn't there be $TARGET instead of example.com in history?

**Comment by ShutdownRepo** [comment]
It should probably be "$DOMAIN" yeah
Not for swaks history though, since the argument is an email address

File: sources/zsh/history.d/dnsx
Comment (ShutdownRepo) [inline sources/zsh/history.d/dnsx:3]: here

File: sources/zsh/history.d/shuffledns
Comment (ShutdownRepo) [inline sources/zsh/history.d/shuffledns:4]: here

File: sources/zsh/history.d/subfinder
Comment (ShutdownRepo) [inline sources/zsh/history.d/subfinder:6]: here

**Review by ShutdownRepo — CHANGES_REQUESTED** [review]
As discussed in previous comments, let's replace the hardcoded domains with `"$DOMAIN"`

**Review by ShutdownRepo — APPROVED** [review]

File: sources/zsh/aliases.d/ligolo-ng
Comment (ShutdownRepo) [inline sources/zsh/aliases.d/ligolo-ng:1]: In my opinion, this shouldn't be an alias, but instead a symbolic link placed in `/opt/tools/bin/`

File: sources/zsh/aliases.d/volatility3
Comment (ShutdownRepo) [inline sources/zsh/aliases.d/volatility3:1]: I'm not sure this alias is needed. I think it could be changed like this: `alias volatility3=vol`

File: sources/zsh/aliases.d/jadx
Comment (ShutdownRepo) [inline sources/zsh/aliases.d/jadx:2]: In my opinion, this shouldn't be aliases, but instead symbolic links placed in `/opt/tools/bin/`

**Review by ShutdownRepo — CHANGES_REQUESTED** [review]

File: sources/zsh/aliases.d/volatility3
Comment (QU35T-code) [reply sources/zsh/aliases.d/volatility3:1]: I put this alias because `vol` is already used by `volatility2`

**Comment by QU35T-code** [comment]
Hey @ShutdownRepo,

I have made the requested changes.
Regarding `volatility`, I let you check what I propose for the alias conflict.

**Review by ShutdownRepo — APPROVED** [review]

**Comment by ShutdownRepo** [comment]
Tests were skipped, workflow should be fixed now

File: sources/zsh/aliases.d/volatility3
Comment (Dramelac) [reply sources/zsh/aliases.d/volatility3]: Or use `vol3` for `volatility3` and keep `vol` for `volatility2` ?

Comment (QU35T-code) [reply sources/zsh/aliases.d/volatility3]: I don't know. It would be clearer to put 'vol2' and 'vol3' ?

**Comment by ShutdownRepo** [comment]
Good to go, merging now

**Pull request merged**
